### ISSUE TITLE
Azure Arc connectedk8s support for arm64 

### ISSUE THREAD

**Issue by mikebranstein** [Azure/azure-cli-extensions#2701]
### Extension name (the extension in question)
connectedk8s

### Description of issue (in as much detail as possible)
Configuration:
- Platform: Raspberry Pi 4
- OS: Ubuntu Server 20.10
- K8s: v1.19.4

I'm trying to enroll my Raspberry Pi 4 K8s cluster in Azure Arc:

`> az connectedk8s connect --name <name> --resource-group <resource group> --location eastus`

The enrollment times out and all pods report `FailedScheduling`:
```
> for pod in $(kubectl get pods -n azure-arc | grep 'Pending' | cut -d' ' -f1); do kubectl describe pod $pod -n azure-arc | grep FailedScheduling; done

  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
  Warning  FailedScheduling  22m   default-scheduler  0/5 nodes are available: 5 node(s) didn't match node selector.
```

Digging into the pod manifest, they are requesting amd64 nodes:

```
  nodeSelector:
    kubernetes.io/arch: amd64
    kubernetes.io/os: linux
```

I'm running on arm64, and nodes a labeled as such (b/c I'm on Raspberry Pi):
```
  labels:
    beta.kubernetes.io/arch: arm64
    beta.kubernetes.io/os: linux
    kubernetes.io/arch: arm64
    kubernetes.io/hostname: rpi4-02
    kubernetes.io/os: linux
```

Is the amd64 node selector an oversight, or intentional? Does Azure Arc support arm64?
-----

**Comment by yonzhan** [comment]
connectedk8s

**Comment by PostlMC** [comment]
Seeing exactly the same thing there, also on K8s 1.19.4 and Ubuntu 20.04, with a cluster of Pi 4s.  The recent Cloud9 video seems to imply that ARM is supported.  Perhaps we can manually edit the chart before deployment for now?

**Comment by mikebranstein** [comment]
That's what I'm thinking @PostlMC. The issue appears to be an assumption that amd64 is the preferred node type. 

I'm not sure if we'll be able to modify the deployment before it has been applied to the cluster, but we may be able to edit it after it's been applied. That should allow the scheduler to select an appropriate node.

**Comment by shashankbarsin** [comment]
Azure Arc enabled K8s doesn't have support for arm64 yet. We currently have this in our backlog and will need to work/collaborate on making the following ready for arm64 - Arc agents, flux and its dependencies. The transitive dependencies may take more time for arm64 readiness. To help us prioritise this in our backlog, it'd be great if we could discuss the scenarios (GitOps, Azure Monitor, Azure Policy,...) that are of immediate priority to you and your adoption timelines (for POC, production workloads,...). @mikebranstein, @PostlMC  please reach out to me on shasb@microsoft.com. Happy to gather requirements and adjust our roadmap and priorities to help you out.

**Comment by mohsinonxrm** [comment]
@shashankbarsin , how was shown to be working in the Azure Friday video? I've seen it as well.

Anyway, I've got it working by building a mixed-arch (arm64 and amd64) k3s (3 RPI 4 and 1 Rock Pi X).

However, I'm struggling to get az appservice kube complete the provisioning. I can see the pods running on amd64 node. However, they fail on the arm64 nodes. Will dig further in the logs and open another issue.

**Comment by brendonts** [comment]
I'm also very interested in ARC support for arm64. This would enable me to test Arc at a small scale so I can better understand its usefulness for clients.

**Comment by MaxiPalle** [comment]
Would be great to integrate arm64 machines. $customer uses 100s of Raspberrys  and would love to use it.

**Comment by sysgazer** [comment]
Trying multipass/k3d/azure arc on my M1 Mac but to no avail. Any idea if the support for arm64 is coming soon? :)

**Comment by kingdonb** [comment]
With my `az` cli installed in the default location, on an aarch64 Linux (arm64) I noticed it was possible to install the connectedk8s extension, until it tried to download and install an amd64 binary for helm and resulted in an `exec format error`.

```
$ cd ~/.azure/helm/v3.6.3/linux-amd64
$ wget https://get.helm.sh/helm-v3.6.3-linux-arm64.tar.gz
$ tar xvf helm-v3.6.3-linux-arm64.tar.gz
$ mv linux-arm64/helm helm
```

My expectations were pretty low that this would work, given the kind cluster behind this node is also arm64, and indeed the images behind Arc are apparently not multiarch either. That's a shame! Arm64 machines are much cheaper.

With this announcement, I think we should prioritize adding arm64 support for the `connectedk8s` extension!

https://azure.microsoft.com/en-us/updates/public-preview-arm64based-azure-vms-can-deliver-up-to-50-better-priceperformance/

**Comment by drcrook1** [comment]
I'm simply trying to do local development using my M1 based Mac Studio and having this same problem.  I will upvote this.

**Comment by tobiasoort** [comment]
Similar. Since moving to M1 and Volterra I have no Intel boxes in my homelab anymore. Would love to tinker with Arc to build out PoCs but am now blocked due to architecture support. If there's opensource stuff I can help with please let me know. I'm also currently getting arm64-support working for Apache Camel-K.

**Comment by XavierGeerinck** [comment]
Also awaiting this! Seeing price / perf ratio we can't do without ARM nodes

**Comment by omkark95** [comment]
Azure Arc-enabled Kubernetes now supports ARM64 nodes: https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/system-requirements#cluster-requirements

**Comment by mohsinonxrm** [comment]
err... @omkark95 , the cluster extensions still aren't supported on arm64: https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/system-requirements#cluster-requirements

**Comment by omkark95** [comment]
@mohsinonxrm that's correct, currently we support feature such as cluster connect and viewing Kubernetes resources in the Azure portal. Work item to support cluster extension on ARM64 is in our backlog. To help us prioritise this in our backlog, it'd be great if we could discuss the scenarios (GitOps, Azure Monitor, Azure Policy, KeyVault,...) that are of immediate priority to you and your adoption timelines. Happy to gather requirements and adjust our roadmap and priorities to help you out.

**Comment by devgig** [comment]
Arm64 is required to evaluate this feature and also for home lab use.

**Comment by Ahmad-Bunni** [comment]
I managed to get it working on a jetson which is an arm64 based.

![Image](https://github.com/user-attachments/assets/66588a1f-6e91-47e7-b075-4f921fc2eb75)

Steps:
- Upgrade to latest Azure CLI
- Install connectedk8s extension  `az extension add --name connectedk8s`
- Make sure you have your cluster setup. In my case it is a k3s.
- Generate the command from azure portal to add it as an existing arc cluster
- Export your `KUBECONFIG` as part of `~/.bashrc` or you can specify `--kube-config` path as part of `az connectedk8s connect ` command you generate from azure.

***IMPORTANT*** 

If you run the `az connectedk8s connect` command you generated, it will crash at helm because the one it uses is amd64. 

Here is the work around, we will replace helm with arm64 but keep it under linux-amd64 folder:

```
cd ~/.azure

mkdir -p ~/.azure/helm/v3.12.2

cd ~/.azure/helm/v3.12.2

wget https://get.helm.sh/helm-v3.12.2-linux-arm64.tar.gz

tar -zxvf helm-v3.12.2-linux-arm64.tar.gz

mv linux-arm64 linux-amd64
```

Now you can re-run the `az connectedk8s connect` command that you generated from azure portal to add the cluster and everything should work.

NOTE
At the time of writing this, the helm version it points to is v3.12.2, so in the future you may need to change it to whatever it attempts to use, which will be shown to you when it crashes during `az connectk8s connect`.
